src/coreclr/src/jit/liveness.cpp

```diff
@@ -2109,40 +2109,80 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                 {
                     break;
                 }
-                __fallthrough;
+                fgTryRemoveNonLocal(node, &blockRange);
+            }
+            break;
 
-            default:
-                assert(!node->OperIsLocal());
-                if (!node->IsValue() || node->IsUnusedValue())
+            case GT_BLK:
+            case GT_OBJ:
+            case GT_DYN_BLK:
+            {
+                bool removed = fgTryRemoveNonLocal(node, &blockRange);
+                if (!removed && node->IsUnusedValue())
                 {
-                    // We are only interested in avoiding the removal of nodes with direct side-effects
-                    // (as opposed to side effects of their children).
-                    // This default case should never include calls or assignments.
-                    assert(!node->OperRequiresAsgFlag() && !node->OperIs(GT_CALL));
-                    if (!node->gtSetFlags() && !node->OperMayThrow(this))
-                    {
-                        JITDUMP("Removing dead node:\n");
-                        DISPNODE(node);
-
-                        node->VisitOperands([](GenTree* operand) -> GenTree::VisitResult {
-                            operand->SetUnusedValue();
-                            return GenTree::VisitResult::Continue;
-                        });
-
-                        blockRange.Remove(node);
-                    }
+                    // IR doesn't expect dummy uses of `GT_OBJ/BLK/DYN_BLK`.
+                    node->ChangeType(TYP_BYTE);
+                    node->ChangeOper(GT_NULLCHECK);
+                    block->bbFlags |= BBF_HAS_NULLCHECK;
+                    optMethodFlags |= OMF_HAS_NULLCHECK;
+                    JITDUMP("Replace an unused OBJ/BLK node [%06d] with a NULLCHECK\n", dspTreeID(node));
                 }
+            }
+            break;
+
+            default:
+                fgTryRemoveNonLocal(node, &blockRange);
                 break;
         }
     }
 }
 
+//---------------------------------------------------------------------
+// fgTryRemoveNonLocal - try to remove a node if it is unused and has no direct
+//   side effects.
+//
+// Arguments
+//    node       - the non-local node to try;
+//    blockRange - the block range that contains the node.
+//
+// Return value:
+//    None
+//
+// Notes: local nodes are processed independently and not-expected in this funtion.
+//
+bool Compiler::fgTryRemoveNonLocal(GenTree* node, LIR::Range* blockRange)
+{
+    assert(!node->OperIsLocal());
+    if (!node->IsValue() || node->IsUnusedValue())
+    {
+        // We are only interested in avoiding the removal of nodes with direct side-effects
+        // (as opposed to side effects of their children).
+        // This default case should never include calls or assignments.
+        assert(!node->OperRequiresAsgFlag() && !node->OperIs(GT_CALL));
+        if (!node->gtSetFlags() && !node->OperMayThrow(this))
```

Comment:
That condition was pre-existing but it is not always correct,  lower replaces many `HWIntrinsic` and other nodes as `IND` (for example Vector64.Create) without setting `GTF_IND_NONFAULTING` on these `IND` nodes. If you call `fgDebugCheckFlags` on such `IND` that is not marked as `X` and not marked `GTF_IND_NONFAULTING` it would fail.

A better condition would be `if (!node->gtSetFlags() && (!node->OperMayThrow(this)) || ((node->gtFlags & GTF_EXCEPT) == 0)` but such fix doesn't produce many positive diffs, so I would wait for 
https://github.com/dotnet/runtime/issues/10249 to show all such places and then fix the root cause, not the side-effects.

Reply:
I believe in LIR `GTF_EXCEPT` is not supposed to capture children side effects. What happens if you use `if (!node->gtSetFlags()  && ((node->gtFlags & GTF_EXCEPT) == 0)` here?

Reply:
> I believe in LIR GTF_EXCEPT is not supposed to capture children side effects

Yes, but we don't check it and right now many parent nodes are having `GTF_EXCEPT` from children, so with `if (!node->gtSetFlags() && ((node->gtFlags & GTF_EXCEPT) == 0)` we are keeping many null checks that we can delete.

Reply:
When you say "it is not always correct" you mean that it is overly conservative, correct?
It seems to me that the better approach would be to fix the setting of `GTF_IND_NONFAULTING`.

Reply:
> When you say "it is not always correct" you mean that it is overly conservative, correct?

Yes.

> It seems to me that the better approach would be to fix the setting of GTF_IND_NONFAULTING.

Agree, I should have said "a better temporary hack", I think we should set `GTF_IND_NONFAULTING` for 6.0 after  #10249